File: binsifter/core/iocs.py
"""IOC extraction from FLOSS string output - regex-mined IP addresses,
URLs, domains, and registry paths.

Direct port of BinSifter-Rowan_v1.3.0-beta.1.ps1, lines ~2329-2365 (the
v1.3-proto1 IOC mining step) - same four regexes, same case-insensitive
dedup, same 50-item display cap, since this is exactly the kind of "match
the original precisely" logic per project convention (see file_type.py's
docstring for the same rationale).

Deliberately NOT "improved" versus the original even where the original's
regexes have quirks:
  - None of the four patterns carry an IgnoreCase option in the PowerShell
    version ($ipRegex/$urlRegex/$domainRegex/$regRegex are all plain
    [regex] literals with no RegexOptions), so matching is case-sensitive
    exactly as written there. In particular the domain pattern's character
    classes are all lowercase ([a-z0-9]), so it only matches all-lowercase
    domains - an UPPERCASE.COM string in a binary would not be flagged.
    This looks like it could be an oversight in the original, but silently
    "fixing" it here would make this port's output diverge from the
    PowerShell version's for the exact same input file, which is worse for
    side-by-side verification than reproducing a known quirk. Flag it to
    Steve before changing it.
  - Similarly, the URL pattern only matches a lowercase "http"/"https"
    scheme literally (no case-insensitivity), so "HTTPS://..." strings
    aren't matched either. Same reasoning - ported as-is.

Runs on the same FLOSS strings already extracted for PossibleFalseNegative
files (see floss_scan.py) - never triggers a second/separate FLOSS
invocation. Best-effort: mirrors the PowerShell version's inner try/catch
around this step (a bad string or regex edge case must never fail the
file's scan) - callers should not need their own try/except around this,
extract_iocs() itself never raises.

Caveat not present in the original: .NET's regex \b (word boundary) and
Python's re \b are not guaranteed to be byte-for-byte identical on exotic
Unicode input (both define "word character" slightly differently at the
edges) - not expected to matter for the ASCII-oriented patterns here, but
noted since exact-match fidelity is the whole point of this module.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

# Same four patterns as the PowerShell version, translated 1:1 - see
# BinSifter-Rowan_v1.3.0-beta.1.ps1 lines 2349-2352. No re.IGNORECASE on any of
# them, matching the original's lack of a RegexOptions.IgnoreCase - see the
# module docstring for why that's deliberate, not an oversight here.
_IP_RE = re.compile(
    r"\b(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\b"
)
_URL_RE = re.compile(r"""\bhttps?://[^\s"'<>]{4,200}""")
_DOMAIN_RE = re.compile(
    r"\b(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\.)+(?:com|net|org|io|ru|cn|biz|info|xyz|top|club|online|site|tk|cc)\b"
)
_REGISTRY_RE = re.compile(r"""\bHKEY_[A-Z_]+\\[^\s"']{2,200}""")

# Original caps the CSV column at 50 entries so a pathological string blob
# can't balloon a report row - see the PowerShell version's comment at
# line 2364 ("Capped so a pathological string blob can't balloon the CSV
# row.").
_MAX_DISPLAYED_IOCS = 50


@dataclass
class IocExtractionResult:
    count: int
    # "; "-joined, capped at _MAX_DISPLAYED_IOCS entries - "" if none found.
    display: str


def extract_iocs(strings: list[str]) -> IocExtractionResult:
    """Mines a list of already-extracted FLOSS strings for IOC-shaped
    values (IPs, URLs, domains, registry paths).

    Dedup is case-insensitive, matching the PowerShell version's
    HashSet[string] built with StringComparer.OrdinalIgnoreCase - but the
    first-seen casing is what's kept/displayed (same as a real .NET
    HashSet, which stores the first-inserted instance and just treats a
    later case-variant as "already present"), except domain matches, which
    are explicitly lowercased before insertion, same as the original's
    `$m.Value.ToLowerInvariant()`.

    Insertion order (== first-seen order across IP, then URL, then domain,
    then registry-path regexes, per string, in string-list order) is
    preserved for the first 50 entries used in the display join, mirroring
    a real .NET HashSet<T>'s practical (not contractually guaranteed, but
    consistently observed) enumeration order for a set that's only ever
    added to, never removed from - same assumption the original's
    `Select-Object -First 50` relies on.
    """
    seen: dict[str, str] = {}  # lowercased key -> first-seen-cased display value

    def _add(value: str) -> None:
        key = value.lower()
        if key not in seen:
            seen[key] = value

    for s in strings:
        if not s:
            continue
        for m in _IP_RE.finditer(s):
            _add(m.group(0))
        for m in _URL_RE.finditer(s):
            _add(m.group(0))
        for m in _DOMAIN_RE.finditer(s):
            _add(m.group(0).lower())
        for m in _REGISTRY_RE.finditer(s):
            _add(m.group(0))

    count = len(seen)
    if count == 0:
        return IocExtractionResult(count=0, display="")

    display = "; ".join(list(seen.values())[:_MAX_DISPLAYED_IOCS])
    return IocExtractionResult(count=count, display=display)

File: ingot/crates/ingot-core/src/iocs.rs
//! IOC extraction from FLOSS strings - regex-mined IPs, URLs, domains, and
//! registry paths. Direct port of `binsifter.core.iocs`.
//!
//! The four patterns are translated 1:1 from the PowerShell version and are
//! **deliberately not "improved"**, even where they have quirks - matching
//! the other variants' output for the same input matters more than a
//! cleaner regex:
//!
//! * none of the patterns are case-insensitive, so the domain pattern
//!   (lowercase character classes) only matches all-lowercase domains and
//!   the URL pattern only matches a lowercase `http`/`https` scheme.
//! * dedup is case-insensitive but the first-seen casing is kept; domain
//!   matches are lowercased before insertion.
//! * insertion order (IP, then URL, then domain, then registry, per string,
//!   in list order) is preserved for the first 50 entries in the display.

use std::sync::LazyLock;

use regex::Regex;

static IP_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"\b(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\b").unwrap()
});
static URL_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r#"\bhttps?://[^\s"'<>]{4,200}"#).unwrap());
static DOMAIN_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"\b(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\.)+(?:com|net|org|io|ru|cn|biz|info|xyz|top|club|online|site|tk|cc)\b").unwrap()
});
static REGISTRY_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r#"\bHKEY_[A-Z_]+\\[^\s"']{2,200}"#).unwrap());

const MAX_DISPLAYED_IOCS: usize = 50;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct IocExtractionResult {
    pub count: usize,
    /// `"; "`-joined, capped at [`MAX_DISPLAYED_IOCS`]; `""` if none.
    pub display: String,
}

/// Mine already-extracted strings for IOC-shaped values. Never errors.
pub fn extract_iocs<S: AsRef<str>>(strings: &[S]) -> IocExtractionResult {
    // lowercased key -> first-seen-cased display value, insertion-ordered
    let mut seen: Vec<(String, String)> = Vec::new();
    let mut add = |value: String| {
        let key = value.to_lowercase();
        if !seen.iter().any(|(k, _)| *k == key) {
            seen.push((key, value));
        }
    };

    for s in strings {
        let s = s.as_ref();
        if s.is_empty() {
            continue;
        }
        for m in IP_RE.find_iter(s) {
            add(m.as_str().to_string());
        }
        for m in URL_RE.find_iter(s) {
            add(m.as_str().to_string());
        }
        for m in DOMAIN_RE.find_iter(s) {
            add(m.as_str().to_lowercase());
        }
        for m in REGISTRY_RE.find_iter(s) {
            add(m.as_str().to_string());
        }
    }

    let count = seen.len();
    if count == 0 {
        return IocExtractionResult {
            count: 0,
            display: String::new(),
        };
    }
    let display = seen
        .iter()
        .take(MAX_DISPLAYED_IOCS)
        .map(|(_, v)| v.as_str())
        .collect::<Vec<_>>()
        .join("; ");
    IocExtractionResult { count, display }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn extracts_all_kinds_in_first_seen_order() {
        let strings = vec![
            "connect to 192.168.1.5 then GET https://evil.example.com/payload".to_string(),
            "reg key HKEY_LOCAL_MACHINE\\Software\\Run\\x also c2.bad.top here".to_string(),
        ];
        let r = extract_iocs(&strings);
        // string 1: IP, URL, domain(from the URL host); string 2: domain, registry
        assert_eq!(r.count, 5);
        let parts: Vec<&str> = r.display.split("; ").collect();
        assert_eq!(parts[0], "192.168.1.5");
        assert_eq!(parts[1], "https://evil.example.com/payload");
        assert_eq!(parts[2], "evil.example.com");
        assert_eq!(parts[3], "c2.bad.top");
        assert_eq!(parts[4], "HKEY_LOCAL_MACHINE\\Software\\Run\\x");
    }

    #[test]
    fn case_insensitive_dedup_keeps_first_casing() {
        let s = vec![
            "HTTP is not matched but http://Site.example.io/A".to_string(),
            "again http://site.example.io/a".to_string(),
        ];
        let r = extract_iocs(&s);
        // the two URLs differ only in case -> one entry, first casing kept
        // (but each also yields a domain match, lowercased)
        assert!(r.display.contains("http://Site.example.io/A"));
        assert!(!r.display.contains("http://site.example.io/a"));
    }

    #[test]
    fn uppercase_domain_and_scheme_not_matched_quirk() {
        let r = extract_iocs(&["visit HTTPS://BAD.COM now".to_string()]);
        assert_eq!(
            r.count, 0,
            "uppercase scheme/domain deliberately not matched"
        );
    }

    #[test]
    fn empty_and_none() {
        assert_eq!(extract_iocs::<String>(&[]).count, 0);
        assert_eq!(extract_iocs(&["nothing here".to_string()]).display, "");
    }

    #[test]
    fn display_capped_at_50() {
        let s: Vec<String> = (0..80).map(|i| format!("10.0.0.{i}")).collect();
        let r = extract_iocs(&s);
        assert_eq!(r.count, 80);
        assert_eq!(r.display.split("; ").count(), 50);
    }
}
